### ISSUE TITLE
Normal mode: pressing 'x' to delete char displays clipboard contents, obscures actual text

### ISSUE THREAD

**Issue by brett-mahar** [fox0430/moe#1377]
Moe version: latest git pull as of 24-June-2021

After deleting text, the clipboard content remains on screen, even when entering "insert" mode, makes it difficult to edit text.

This occurs with:

[Clipboard]
enable = true or false

If I delete the entire section of [Standard] then this does not happen. 

nim -v
Nim Compiler Version 1.5.1 [Linux: arm64]
Compiled at 2021-06-17
Copyright (c) 2006-2021 by Andreas Rumpf
git hash: c83ac16671170e513fa1ca40419892ccc618ec85
active boot switches: -d:release

**Comment by fox0430** [comment]
@brett-mahar 

Thank you for your bug report.

Could you tell me more about that? (Picture, operation, etc...)

**Comment by brett-mahar** [comment]
![2021-06-26___06:56:29](https://user-images.githubusercontent.com/82318539/123484574-ebaff280-d64b-11eb-9b81-67453d2cd7d3.png)

Hi @fox0430 

I attached a screenshot above, the clipboard contained "instance spring" and while in "normal mode" I pressed 'x' once, the contents of the clipboard obscured the text (appeared in orange). This didn't happen with Moe build a week or two ago (following the fix you made from my other bug report).

I tried playing with the config but multiple different options in the [Standard] part of the file seemed to trigger this (ie I could not just delete one option to stop this happening).

I also tried to set as many options as possible to false but that also did not stop this happening.

Let me know if you want more info.
My config:

[Standard]
theme = "dark"
number = false
currentNumber = true
cursorLine = false
statusLine = true
tabLine = false
syntax = true
indentationLines = false
tabStop = 2
autoCloseParen = false
autoIndent = false
ignorecase = true
smartcase = true
disableChangeCursor = false
defaultCursor = "noneBlinkBlock"
normalModeCursor = "noneBlinkBlock"
insertModeCursor = "blinkIbeam"
autoSave = false
liveReloadOfConf = false
incrementalSearch = true
popUpWindowInExmode = false
autoDeleteParen = false
smoothScroll = false

[Clipboard]
enable = true

[BuildOnSave]
enable = false

[TabLine]
allBuffer = true

[StatusBar]
multipleStatusBar = false
merge = false
mode = true
filename = false
chanedMark = false
line = true
column = false
encoding = false
language = false
directory = false
gitbranchName = false
showGitInactive = false
showModeInactive = false

[Highlight]
currentLine = true
reservedWord = ["TODO", "WIP", "NOTE"]
replaceText = false
pairOfParen = true
fullWidthSpace = true
trailingSpaces = true
currentWord = false

[AutoBackup]
enable = false

[QuickRun]
saveBufferWhenQuickRun = false

[Notification]
screenNotifications = true
logNotifications = false
autoBackupScreenNotify = false
autoBackupLogNotify = false
autoSaveScreenNotify = false
autoSaveLogNotify = false
yankScreenNotify = true
yankLogNotify = false
deleteScreenNotify = true
deleteLogNotify = false
saveScreenNotify = true
saveLogNotify = false
quickRunScreenNotify = false
quickRunLogNotify = false
buildOnSaveScreenNotify = false
buildOnSaveLogNotify = false
filerScreenNotify = true
filerLogNotify = false
restoreScreenNotify = true
restoreLogNotify = false

[Filer]
showIcons = true

[Autocomplete]
enable = true

[Persist]
exCommand = true
search = true
cursorPosition = true

#    # `release` build gives errors if below `debug` lines aren't commented out:
#[Debug.WindowNode]
#enable = false

#[Debug.EditorView]
#enable = false

#[Debug.BufferStatus]
#enable = false

[Theme]
#baseTheme = "vscode"

editorBg = "default"
lineNum = "gray54"
lineNumBg = "default"
currentLineNum = "teal"
currentLineNumBg = "default"

statusLineNormalMode = "white"
statusLineNormalModeBg = "darkOrange3_1"
statusLineModeNormalMode = "black"
statusLineModeNormalModeBg = "white"
statusLineNormalModeInactive = "darkOrange3_1"
statusLineNormalModeInactiveBg = "white"

statusLineInsertMode = "white"
statusLineInsertModeBg = "darkOrange3_1"
statusLineModeInsertMode = "black"
statusLineModeInsertModeBg = "white"
statusLineInsertModeInactive = "darkOrange3_1"
statusLineInsertModeInactiveBg = "white"

statusLineVisualMode = "white"
statusLineVisualModeBg = "darkOrange3_1"
statusLineModeVisualMode = "black"
statusLineModeVisualModeBg = "white"
statusLineVisualModeInactive = "darkOrange3_1"
statusLineVisualModeInactiveBg = "white"

statusLineReplaceMode = "white"
statusLineReplaceModeBg = "darkOrange3_1"
statusLineModeReplaceMode = "black"
statusLineModeReplaceModeBg = "white"
statusLineReplaceModeInactive = "darkOrange3_1"
statusLineReplaceModeInactiveBg = "white"

statusLineFilerMode = "white"
statusLineFilerModeBg = "darkOrange3_1"
statusLineModeFilerMode = "black"
statusLineModeFilerModeBg = "white"
statusLineFilerModeInactive = "darkOrange3_1"
statusLineFilerModeInactiveBg = "white"

statusLineExMode = "white"
statusLineExModeBg = "darkOrange3_1"
statusLineModeExMode = "black"
statusLineModeExModeBg = "white"
statusLineExModeInactive = "darkOrange3_1"
statusLineExModeInactiveBg = "white"

statusLineGitBranch = "white"
statusLineGitBranchBg = "darkOrange3_1"

# tab line
tab = "white"
tabBg = "default"
currentTab = "white"
currentTabBg = "darkOrange3_1"

# command  bar
commandBar = "gray100"
commandBarBg = "default"

# error message
errorMessage = "red"
errorMessageBg = "default"

# search result highlighting
searchResult = "black"
searchResultBg = "seaGreen1_2"

# selected area in visual mode
visualMode = "gray100"
visualModeBg = "purple_1"

# color scheme
defaultChar = "gray84"
#defaultChar = "lightGoldenrod2_2"
gtKeyword = "lightGoldenrod2_2"
gtFunctionName = "gold1"
gtBoolean = "yellow"
gtStringLit = "yellow"
gtSpecialVar = "green"
gtBuiltin = "yellow"
gtDecNumber = "aqua"
gtComment = "seaGreen1_2"
gtLongComment = "seaGreen1_2"
gtWhitespace = "gray"
gtPreprocessor = "green"

# filer mode
currentFile = "gray100"
currentFileBg = "teal"
file = "gray100"
fileBg = "default"
dir = "darkOrange3_1"
dirBg = "default"
pcLink = "teal"
pcLinkBg = "default"

# pop up window
popUpWindow = "gray100"
popUpWindowBg = "black"
popUpWinCurrentLine = "darkOrange3_1"
popUpWinCurrentLineBg = "black"

# replace text highlighting
replaceText = "default"
replaceTextBg = "red"

# pair of paren highlighting
parenText = "default"
parenTextBg = "darkOrange3_1"

# highlight other uses current word
currentWord = "red"
currentWordBg = "black"

# highlight full width space
#highlightFullWidthSpace = "darkOrange3_1"  
#highlightFullWidthSpaceBg = "teal"

# highlight trailing spaces
#highlightTrailingSpaces = "red"
highlightTrailingSpacesBg = "teal"

# highlight reserved words
reservedWord = "white"
reservedWordBg = "gray"

# configuration mode
currentSetting = "gray100"
currentSettingBg = "teal"

# highlight current line background
currentLineBg = "darkSeaGreen4_1"
#currentLineBg = "darkSeaGreen3_1"

**Comment by fox0430** [comment]
@brett-mahar 

I tried it but couldn't reproduce it.
Can you tell me about your environment? (OS, Terminal emulator, etc...)

**Comment by brett-mahar** [comment]
Hi @fox0430 

Its a PinebookPro running Manjaro Linux (Arch Linux) 5.12.13-1-MANJARO-ARM (testing branch)

I tried it with 2 Terminal emulators: Sakura and ST, and the same thing happens with both.

I copy and paste using the Terminal hotkeys Ctrl+Shift+c Ctrl+Shift+v

My window manager is Sway and I run GTK apps under Wayland. I don't think that could cause this but thought I should mention it.

Brett.

**Comment by fox0430** [comment]
@brett-mahar 

Thank you very much!
Is it possible to tell me the operations that can be reproduced?

For example,

1. Open a file
```
moe test.txt
```
2. Delete character
Write text
```
hello
```
3. Delete a character
Press ```x``` key

**Comment by brett-mahar** [comment]
Hi @fox0430 

1. Open the file
2. Press 'x' to delete a character
3. Whatever is in the clipboard will appear to overwrite the text 

(but if I save the file, that clipboard text isn't there - it is just the words that were part of the original file, minus whichever letter I deleted with 'x' - just like in the screenshot above - the orange text "instance spring" was in the clipboard when I pressed 'x')

Strange mystery that its only happening to me!

**Comment by fox0430** [comment]
@brett-mahar 

Thank you very much!

Could you try another terminal emulator? (urxvt, xterm or alacritty)

**Comment by brett-mahar** [comment]
@fox0430 

I tried xterm, garlicterm, and xfce4-terminal but get the same strange behaviour in all of them.

**Comment by fox0430** [comment]
@brett-mahar 

Ok, thank you.
I'll reproduce your environment and try it out.

**Comment by brett-mahar** [comment]
@fox0430 

I tried a few more things to try to narrow it down:

1. Used moe (as root and also using my normal user account) from a virtual terminal: the problem does not happen then.

2. Used i3/X11 instead of Sway/Wayland as the window manager: get a similar problem, but when I hit 'x', instead of the contents of the clipboard, I get a message "could not start a wayland server" (or something like that, I'm logged back into Sway now, I can't recall the exact words). Even after I selected and copied some text, when I pressed 'x' I always got that message under i3.

3. Tried to reproduce the problem using vim and nvi. They both still work as expected.

**Comment by fox0430** [comment]
@brett-mahar 

Does this bug happen with another command?  (```dw```, ```dd```, etc..,)

**Comment by brett-mahar** [comment]
@fox0430 

It does not happen with `dd`

It does happen with `yy` and `dw` (I don't use those shortcuts much, didn't notice that before).

Another twist: as I tried this, I didn't get the copy of the clipboard appearing in the moe terminal display, but instead got some text from this github page (see screenshot). I hadn't copied anything from this page. Very strange! 

PS I am happy to keep trying things to try and solve this, but if you don't want to fix a bug that only seems to affect me, I totally understand :-)

![2021-06-29___20:14:48](https://user-images.githubusercontent.com/82318539/123780946-0c5f9d00-d917-11eb-9c56-7fdbf8df614a.png)

**Comment by fox0430** [comment]
@brett-mahar 

It's very strange. I want to know this problem cause. And, I installed Manjaro (GNOME 21.0.7) and tried it. But, works fine.

I set it up with the following command.

```
sudo pacman -S base-devel
curl https://nim-lang.org/choosenim/init.sh -sSf | sh
git clone https://github.com/fox0430/moe
cd moe
nimble build
```

Is there something missing or different?

**Comment by brett-mahar** [comment]
Hi @fox0430 

I have base-devel installed, but I didn't use choosenim. To update Nim, I do:
cd /home/brett/.NIM/Nim
git pull
./build_all.sh

I built moe the same as you describe above. I have tried using clang and gcc as backend but still see the bug (i.e. in /etc/nim/nim.cfg tried both cc = clang or cc = gcc). I also tried `nimble release` but still got the bug.

Did you install Manjaro on an x64 computer?

Maybe it is the combination of PinebookPro (which uses an ARM cpu), Nim, and Moe? Triggering some strange memory leak or something? I don't know!

**Comment by brett-mahar** [comment]
@fox0430 

I just rebuilt Nim (git pull used to update source to latest devel version) and then used it to rebuild Moe (updated with git pull), since I saw you have added some commits to the editor.

Unfortunately the problem is still there.

Only new info I have, is when the clipboard is empty (like now I hadn't copied anything yet while in the terminal using ctrl+shift+c since last turning on the computer), instead of 'x' showing clipboard contents, it shows 'No selection'

Just mentioning it in case it is a clue :-)

**Comment by fox0430** [comment]
@brett-mahar 

Ok, Thank you for your additional information.
I try it on x86_64 computer (Ryzen 9 5900x).

I'll think about something a little more.

Does this problem occur in moe v0.2.6.1?

**Comment by brett-mahar** [comment]
@fox0430 

I just installed moe v0.2.6.1 with `nimble install moe` but I still see the same bug.

**Comment by fox0430** [comment]
@brett-mahar 

Ok, thank you.

What clipboard tool are you using? xclip?

Can I also try moe v0.1.0? (before moe supported clipboard)

**Comment by brett-mahar** [comment]
@fox0430 

I use wl-clipboard (https://aur.archlinux.org/packages/wl-clipboard-git/) and wl-clipboard-x11 (https://aur.archlinux.org/packages/wl-clipboard-x11/) which imitates the x11 tool xsel.

To rule them out I uninstalled them both and rebooted. But it looks like the same bug, just causing a different effect (see screenshot):

![error](https://user-images.githubusercontent.com/82318539/124260419-04039e00-db73-11eb-9c1f-aee6e3e01435.png)

I don't know how to get the code for the old version v0.1.0. Is there a git command for that?

**Comment by fox0430** [comment]
@brett-mahar 

Thank you. 

It may not be relevant but I will support wl-clipboard.

And, to use moe v0.1.0 please try the following commands.
```
cd moe
git switch -c v0.1.0 refs/tags/v0.1.0
nimble build
moe -v
```

**Comment by brett-mahar** [comment]
@fox0430 

This problem does not occur with moe v0.1.0 :-)

**Comment by brett-mahar** [comment]
Hi @fox0430 

I just did `nimble build` on the develop branch of Moe.

With 
[Clipboard]
enable = false

in the config, I no longer get the bug with that version.

If I set 
[Clipboard]
enable = true

then the bug returns (previously I would get the bug if this was set to true or false).

Brett.

**Comment by fox0430** [comment]
@brett-mahar 

I added support for wl-clipboard.
Could you try the latest develop branch?
And, Please use this setting.

```
[Clipboard]
enable = true
toolOnLinux = "wl-clipboard"
```

**Comment by brett-mahar** [comment]
That has fixed it! Rebuilding and using toolOnLinux = "wl-clipboard" I no longer have the bug.

Thanks @fox0430 :-)

**Comment by fox0430** [comment]
@brett-mahar 

Great!
Probably, this issue is due to failed clipboard command and show the error message in moe view.
But, I don't know why the unrelated text is showing up.

Does this bug still occur when xclip or xsel set?

**Comment by brett-mahar** [comment]
Hi @fox0430 

I tried the devel version again and it looks like you've fixed it, with:

[Clipboard]
enable = false

I don't get the bug (I did before).

Also (I'm a bit embarrased to say), I realised that when I installed wl-clipboard-x11 (wrapper for wl-clipboard so it will impersonate xsel), pacman uninstalled xsel (they conflict).
So when I was testing before with:

[Clipboard]
enable = true
toolOnLinux = "xsel"

it was triggering this behaviour/bug, but now xsel is installed, I don't see the bug using this config either. Sorry about that!

So the bug no longer occurs for me with any config options, I think it can be closed.

Brett.
